cloudevents/working-drafts/dds-format.md

```diff
@@ -0,0 +1,247 @@
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][rfc2119].
+
+### 1.2 Content-Type
+
+There is no official IANA *media-type* designation for DDS, as such this
+specification uses 'application/dds' to identify such content.
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are represented in the DDS
+[schema][dds-schema].
+
+## 2.1 Type System
+
+The CloudEvents type system is mapped to IDL datatypes as follows :
+
+| CloudEvents   | DDS |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][dds-schema] |
+| Integer       | [int32][dds-schema] |
+| String        | [string][dds-schema] |
+| Binary        | [bytes][dds-schema] |
+| URI           | [string][dds-schema] following [RFC 3986 §4.3][rfc3986-section43]|
+| URI-reference | [string][dds-schema] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [Timestamp][dds-schema]  |
+
+## 2.3 REQUIRED Attributes
+
+REQUIRED attributes are represented explicitly as fields in the [Event type definition][dds-schema].
+
+## 2.4 OPTIONAL Attributes & Extensions
+
+OPTIONAL and extension attributes are represented using a sequence of key-value
+constructs (DDS-CE-Attributes) enabling direct support of the CloudEvent
+[type system][ce-types]. An underlying union and enumeration data structure
+are needed to construct the key-value sequence to establish the association
+between the IDL types and the CloudEvent type system.
+
+```xml
+
+  <enum name="AttributeType">
+   <enumerator name="BOOL"/>
+   <enumerator name="INT32"/>
+   <enumerator name="STRING"/>
+   <enumerator name="BYTES"/>
+   <enumerator name="URI"/>
+   <enumerator name="URI_REF"/>
+   <enumerator name="TIMESTAMP"/>
+ </enum>
+	  
+ <union name="AttributeValue" extensibility="final">
+   <discriminator type="nonBasic" nonBasicTypeName="io::cloudevents::AttributeType"/>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BOOL)"/>
+     <member name="ce_boolean" type="boolean"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::INT32)"/>
+     <member name="ce_integer" type="int32"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::STRING)"/>
+     <member name="ce_string" type="string" stringMaxLength="255"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::BYTES)"/>
+     <member name="ce_bytes" type="byte" sequenceMaxLength="100"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::URI)"/>
+     <member name="ce_uri" type="nonBasic" nonBasicTypeName="io::cloudevents::URI"/>
+   </case>
+   <case>
+     <caseDiscriminator value="(io::cloudevents::URI_REF)"/>
+     <member name="ce_uri_reference" type="nonBasic" nonBasicTypeName="io::cloudevents::URI_Ref"/>
+   </case>
+ </union>
+
+ <struct name="DDS-CE-Attribute">
+   <member name="key" type="string" stringMaxLength="255"/>
```

Comment:
The spec says attribute names SHOULD NOT exceed 20 chars.... notice it's "SHOULD NOT", not "MUST NOT", so the 255 is probably ok here, but I wonder if we should add some text to enforce the "SHOULD NOT" aspect... something like: the above definition mandates that attribute names are not longer than 255 characters, however, per the CE spec they SHOULD NOT be longer than 20 characters.  Maybe? BTW, where did "255" come from? Is that a DDS thing?

Reply:
255 is the default max string length for DDS. I will update max length to be 20 characters for attribute name - I don't see any good reason to extend a DDS attribute name longer than that.

Reply:
Actually, I just removed the Attribute name altogether in light of the earlier comment.